Feature/premakedeps

Ohjurot opened the pull request:
**Summary**
Changelog: Feature / Bugfix: Adding support for premake5 (PremakeDeps) fixes #13275 
Docs: https://github.com/conan-io/docs/pull/XXXX

**Features**
- Implements lua file generation for premake consumption (New feature for 2.0)
- Supports multi build with changing architecture and configuration (New for the premake tool)
- Generates a lua file with variables (includedirs, libdirs, ...) per configuration (Based on existing implementation, adapted for multi build and new architecture)
- Generates a lua file with functions (Just build/compiler setup, linker setup and both) for automatic premake project setup per package and profile/config (Breaks compatibility with conan1 but is way more flexible)
- Generates a lua file with functions (see previous) for automatic premake project setup per package
- Generates a global `conandeps.premake5.lua` file with functions (see previous) for automatic premake project setup. 
- Has fallback capabilities to `release` when no fitting conan configuration is available

**Status**
WIP-Status: Code works and has been covered with a basic unit test. Documentation is still required.

**Checklist**
- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

Ohjurot committed:
Step 1: Generate premake files per package config

Ohjurot committed:
Step 2: Produce files per package and global

Ohjurot committed:
Documentation & Fixed typo

Ohjurot committed:
Make the style better

Ohjurot committed:
Merge remote-tracking branch 'upstream/develop2' into feature/premakedeps

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=13350) <br/>All committers have signed the CLA.

Ohjurot committed:
- Added unittest
- Fixed bugs that only occur in "fresh" deployments
- Fixed bugs that only occur in edge cases

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -37,74 +109,200 @@ def __init__(self, deps_cpp_info):
+        pkg_files = []
+        dep_names = []
+        for require, dep in list(host_req.items()) + list(test_req.items()) + list(build_req.items()):
+            dep_name = require.ref.name
```
This ``for`` internals is a bit too long, might benefit from some splitting.

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -37,74 +109,200 @@ def __init__(self, deps_cpp_info):
+            f"{self.tab * indent}{line}" for line in list(filter(None, string.splitlines()))
+        ])
+
+    def _premake_filtered(self, content, configuration, architecture, indent=0):
```
These filters are quite unknown for others (like us) not knowing Premake. A short explanation or a link to somewhere that defines what are those would be nice.

memsharded reviewed (commented):
I think this looks quite good, clean, with templates, new test, etc, good job.

If we want to move this forward for the next patches releases 2.0.X, this PR should be targeted to the ``release/2.0`` branch. As long as we don't document it, it should be very doable.

At some point we might want to add some functional test that really builds something simple, but that can wait a bit if you are running functional testing on your side.

Ohjurot commented:
@memsharded Thank you for your first feedback. I will close this PR then and open a new one based on the correct branch and targeted towards `release/2.0`.
